File: results/0.25/HTTP-1.1/http-1.1-claude-3-5-haiku-20241022.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <hammer/hammer.h>

typedef struct {
    HParser* method;
    HParser* uri;
    HParser* version;
    HParser* headers;
    HParser* body;
} HTTPRequest;

typedef struct {
    HParser* status_code;
    HParser* status_text;
    HParser* version;
    HParser* headers;
    HParser* body;
} HTTPResponse;

HParser* http_method() {
    return h_choice(
        h_literal_str("GET"),
        h_literal_str("POST"),
        h_literal_str("PUT"),
        h_literal_str("DELETE"),
        h_literal_str("HEAD"),
        h_literal_str("OPTIONS"),
        h_literal_str("TRACE"),
        h_literal_str("CONNECT"),
        h_literal_str("PATCH"),
        NULL
    );
}

HParser* http_version() {
    return h_sequence(
        h_literal_str("HTTP/"),
        h_choice(
            h_literal_str("1.0"),
            h_literal_str("1.1"),
            NULL
        ),
        NULL
    );
}

HParser* header_name() {
    return h_many1(h_satisfy_charclass(isprint));
}

HParser* header_value() {
    return h_many1(h_satisfy_charclass(isprint));
}

HParser* http_header() {
    return h_sequence(
        header_name(),
        h_ch(':'),
        h_ch(' '),
        header_value(),
        h_ch('\r'),
        h_ch('\n'),
        NULL
    );
}

HParser* http_headers() {
    return h_many(http_header());
}

HParser* http_request() {
    return h_sequence(
        http_method(),
        h_whitespace(h_ch(' ')),
        h_many1(h_satisfy_charclass(isprint)),
        h_whitespace(h_ch(' ')),
        http_version(),
        h_ch('\r'),
        h_ch('\n'),
        http_headers(),
        h_ch('\r'),
        h_ch('\n'),
        h_optional(h_many1(h_satisfy_charclass(isprint))),
        NULL
    );
}

HParser* http_response() {
    return h_sequence(
        http_version(),
        h_whitespace(h_ch(' ')),
        h_many1(h_satisfy_charclass(isdigit)),
        h_whitespace(h_ch(' ')),
        h_many1(h_satisfy_charclass(isprint)),
        h_ch('\r'),
        h_ch('\n'),
        http_headers(),
        h_ch('\r'),
        h_ch('\n'),
        h_optional(h_many1(h_satisfy_charclass(isprint))),
        NULL
    );
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <http_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t* buffer = malloc(file_size);
    if (!buffer) {
        perror("Memory allocation error");
        fclose(file);
        return 1;
    }

    size_t read_size = fread(buffer, 1, file_size, file);
    fclose(file);

    if (read_size != file_size) {
        perror("File read error");
        free(buffer);
        return 1;
    }

    HParser* request_parser = http_request();
    HParser* response_parser = http_response();

    HParseResult* request_result = h_parse(request_parser, buffer, read_size);
    HParseResult* response_result = h_parse(response_parser, buffer, read_size);

    if (request_result && request_result->ast) {
        printf("Valid HTTP Request\n");
    } else if (response_result && response_result->ast) {
        printf("Valid HTTP Response\n");
    } else {
        printf("Invalid HTTP Message\n");
    }

    h_parse_result_free(request_result);
    h_parse_result_free(response_result);
    h_destroy(request_parser);
    h_destroy(response_parser);
    free(buffer);

    return 0;
}

File: results/0.25/HTTP-1.1/http-1.1-claude-3-5-haiku-20241022.rs
use nom::{
    branch::alt,
    bytes::complete::{tag, take_while1},
    character::complete::{char, digit1, multispace0},
    combinator::{map, opt},
    multi::{many0},
    sequence::{terminated, tuple},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
struct HttpRequest {
    method: String,
    uri: String,
    version: String,
    headers: Vec<HttpHeader>,
    body: Option<Vec<u8>>,
}

#[derive(Debug)]
struct HttpResponse {
    version: String,
    status_code: u16,
    status_text: String,
    headers: Vec<HttpHeader>,
    body: Option<Vec<u8>>,
}

#[derive(Debug)]
struct HttpHeader {
    name: String,
    value: String,
}

fn parse_method(input: &[u8]) -> IResult<&[u8], String> {
    alt((
        map(tag(b"GET"), |_| "GET".to_string()),
        map(tag(b"POST"), |_| "POST".to_string()),
        map(tag(b"PUT"), |_| "PUT".to_string()),
        map(tag(b"DELETE"), |_| "DELETE".to_string()),
        map(tag(b"HEAD"), |_| "HEAD".to_string()),
        map(tag(b"OPTIONS"), |_| "OPTIONS".to_string()),
        map(tag(b"TRACE"), |_| "TRACE".to_string()),
        map(tag(b"CONNECT"), |_| "CONNECT".to_string()),
    ))(input)
}

fn parse_http_version(input: &[u8]) -> IResult<&[u8], String> {
    map(
        tuple((
            tag(b"HTTP/"),
            take_while1(|c: u8| c == b'.' || (c >= b'0' && c <= b'9')),
        )),
        |(_, version)| String::from_utf8_lossy(version).to_string(),
    )(input)
}

fn parse_header_name(input: &[u8]) -> IResult<&[u8], String> {
    map(
        take_while1(|c: u8| c != b':' && c != b'\r' && c != b'\n'),
        |name| String::from_utf8_lossy(name).to_string(),
    )(input)
}

fn parse_header_value(input: &[u8]) -> IResult<&[u8], String> {
    map(
        take_while1(|c: u8| c != b'\r' && c != b'\n'),
        |value| String::from_utf8_lossy(value).to_string(),
    )(input)
}

fn parse_header(input: &[u8]) -> IResult<&[u8], HttpHeader> {
    map(
        tuple((
            terminated(parse_header_name, char(':')),
            multispace0,
            terminated(parse_header_value, tag(b"\r\n")),
        )),
        |(name, _, value)| HttpHeader { name, value },
    )(input)
}

fn parse_request(input: &[u8]) -> IResult<&[u8], HttpRequest> {
    map(
        tuple((
            terminated(parse_method, char(' ')),
            terminated(take_while1(|c: u8| c != b' '), char(' ')),
            terminated(parse_http_version, tag(b"\r\n")),
            many0(parse_header),
            opt(take_while1(|_| true)),
        )),
        |(method, uri, version, headers, body)| HttpRequest {
            method: method.to_string(),
            uri: String::from_utf8_lossy(uri).to_string(),
            version,
            headers,
            body: body.map(|b| b.to_vec()),
        },
    )(input)
}

fn parse_response(input: &[u8]) -> IResult<&[u8], HttpResponse> {
    map(
        tuple((
            terminated(parse_http_version, char(' ')),
            terminated(map(digit1, |s: &[u8]| std::str::from_utf8(s).unwrap().parse::<u16>().unwrap()), char(' ')),
            terminated(take_while1(|c: u8| c != b'\r' && c != b'\n'), tag(b"\r\n")),
            many0(parse_header),
            opt(take_while1(|_| true)),
        )),
        |(version, status_code, status_text, headers, body)| HttpResponse {
            version,
            status_code,
            status_text: String::from_utf8_lossy(status_text).to_string(),
            headers,
            body: body.map(|b| b.to_vec()),
        },
    )(input)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <input_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_request(&buffer) {
        Ok((_, request)) => {
            println!("Parsed Request: {:?}", request);
        }
        Err(_) => {
            match parse_response(&buffer) {
                Ok((_, response)) => {
                    println!("Parsed Response: {:?}", response);
                }
                Err(e) => {
                    eprintln!("Failed to parse HTTP message: {:?}", e);
                    std::process::exit(1);
                }
            }
        }
    }

    Ok(())
}